compiler/rustc_lint/src/builtin.rs

```diff
@@ -1465,6 +1466,10 @@ impl<'tcx> LateLintPass<'tcx> for TypeAliasBounds {
             // Bounds are respected for `type X = impl Trait`
             return;
         }
+        if cx.tcx.type_of(item.owner_id).skip_binder().has_inherent_projections() {
```

Comment:
hmm... why?

Reply:
Why this deep check for inherent projections? Well, bounds on type aliases *are* respected as matter of fact if the RHS contains an IAT since we look at the `ParamEnv` of the type alias during selection in `astconv` to successfully construct a `Ty`. If I didn't do this check here, the lint would misfire / claim a falsehood.

In "IAT, take 3" that we briefly talked about, with a hypothetical `AliasKind::InherentCandidate` we could delay selection and probably wouldn't need to respect the bounds of type aliases.

Note that the `AliasKind::Weak` PR does basically the same thing. Outside of that, ty alias bounds are already respected for opaques, e.g. `type A<T: B> = impl Tr<T>`, so it's nothing too controversial (?).